src/Lean/Environment.lean

```diff
@@ -731,73 +733,77 @@ def throwAlreadyImported (s : ImportState) (const2ModIdx : HashMap Name ModuleId
   withImporting do
-    let (_, s) ← importMods imports |>.run {}
-    let mut numConsts := 0
-    for mod in s.moduleData do
-      numConsts := numConsts + mod.constants.size + mod.extraConstNames.size
-    let mut modIdx : Nat := 0
-    let mut const2ModIdx : HashMap Name ModuleIdx := mkHashMap (capacity := numConsts)
-    let mut constantMap : HashMap Name ConstantInfo := mkHashMap (capacity := numConsts)
-    for mod in s.moduleData do
-      for cname in mod.constNames, cinfo in mod.constants do
-        match constantMap.insert' cname cinfo with
-        | (constantMap', replaced) =>
-          constantMap := constantMap'
-          if replaced then
-            throwAlreadyImported s const2ModIdx modIdx cname
-        const2ModIdx := const2ModIdx.insert cname modIdx
-      for cname in mod.extraConstNames do
-        const2ModIdx := const2ModIdx.insert cname modIdx
-      modIdx := modIdx + 1
-    let constants : ConstMap := SMap.fromHashMap constantMap false
-    let exts ← mkInitialExtensionStates
-    let env : Environment := {
-      const2ModIdx    := const2ModIdx
-      constants       := constants
-      extraConstNames := {}
-      extensions      := exts
-      header          := {
-        quotInit     := !imports.isEmpty -- We assume `core.lean` initializes quotient module
-        trustLevel   := trustLevel
-        imports      := imports.toArray
-        regions      := s.regions
-        moduleNames  := s.moduleNames
-        moduleData   := s.moduleData
-      }
-    }
-    let env ← setImportedEntries env s.moduleData
-    let env ← finalizePersistentExtensions env s.moduleData opts
-    pure env
-where
-  importMods : List Import → StateRefT ImportState IO Unit
-  | []    => pure ()
-  | i::is => do
-    if i.runtimeOnly || (← get).moduleNameSet.contains i.module then
-      importMods is
-    else do
-      modify fun s => { s with moduleNameSet := s.moduleNameSet.insert i.module }
-      let mFile ← findOLean i.module
-      unless (← mFile.pathExists) do
-        throw <| IO.userError s!"object file '{mFile}' of module {i.module} does not exist"
-      let (mod, region) ← readModuleData mFile
-      importMods mod.imports.toList
-      modify fun s => { s with
-        moduleData  := s.moduleData.push mod
-        regions     := s.regions.push region
-        moduleNames := s.moduleNames.push i.module
-      }
-      importMods is
+    let (_, s) ← importModulesCore imports |>.run
+    finalizeImport s imports opts trustLevel
 
 /--
   Create environment object from imports and free compacted regions after calling `act`. No live references to the
   environment object or imported objects may exist after `act` finishes. -/
-unsafe def withImportModules {α : Type} (imports : List Import) (opts : Options) (trustLevel : UInt32 := 0) (x : Environment → IO α) : IO α := do
+unsafe def withImportModules {α : Type} (imports : Array Import) (opts : Options) (trustLevel : UInt32 := 0) (x : Environment → IO α) : IO α := do
```

Comment:
This is used in downstream projects, including Mathlib, so probably this signature change warrants recording in `RELEASES.md`.

Reply:
(The existence of `lakefile.olean` should be noted too, I guess, as projects will need to .gitignore it.)

Reply:
@semorrison Are we going back to updating `RELEASE.md` for most changes? That seemed to have been abandoned earlier this year. (There are a lot of missing breaking changes from the last few months in `v4.0.0` for instance.) If so, it should also be more generally updated with some of key changes to Lake (i.e., the bump to v5.0.0 and merge with core along with this suite of performance fixes and the `lake update` change). As such, it may be better to have a separate PR with a general update.

Reply:
Let's talk about this on zulip first, and sort out a plan.

Reply:
@semorrison Do you think the PR should wait until that is decided or would the resolution come in a follow-up PR?